lib/gooddata/models/data_product.rb

```diff
@@ -0,0 +1,183 @@
+# encoding: UTF-8
+#
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../rest/resource'
+
+module GoodData
+  class DataProduct < Rest::Resource
+    include Mixin::UriGetter
+
+    attr_accessor :domain
+
+    ALL_DATA_PRODUCTS_PATH = '/gdc/domains/%{domain_name}/dataproducts'
+    ONE_DATA_PRODUCT_PATH = '/gdc/domains/%{domain_name}/dataproducts/%{id}'
+
+    #
+    ProvisioningResult = Struct.new('ProvisioningResult', :id, :status, :project_uri, :error)
+
+    class << self
+      def [](id, opts)
+        domain = opts[:domain]
+        fail ArgumentError, 'No :domain specified' if domain.nil?
+
+        client = domain.client
+        fail ArgumentError, 'No client specified' if client.nil?
+
+        if id == :all
+          GoodData::DataProduct.all(opts)
+        else
+          data_products_uri = ONE_DATA_PRODUCT_PATH % { domain_name: domain.name, id: id }
+
+          result = client.get(data_products_uri)
+          client.create(GoodData::DataProduct, result.merge('domain' => domain))
+        end
+      end
+
+      def all(opts = {})
+        domain = opts[:domain]
+        fail ArgumentError, 'No :domain specified' if domain.nil?
+
+        client = domain.client
+        fail ArgumentError, 'No client specified' if client.nil?
```

Comment:
Refers to 'client' more than self (maybe move it to another class?) (FeatureEnvy)